[CUSPARSE] Fix conversions between CuSparseMatrixCOO and CuSparseMatrixCSC

amontoison opened the pull request:
- Fix CuSparseMatrixCOO(csc);
- Add BSR <-> CSC conversions;
- Add tests for conversions between sparse matrices;
- Test sort routines added in # 1637  (`sort_csc`, `sort_csr` and `sort_coo`).

amontoison committed:
[CUSPARSE] Fix conversions between CuSparseMatrixCOO and CuSparseMatrixCSC

maleadt commented:
CI failures related. Note that our wrappers have been generated for CUDA 11.8, so not all functions may be available on older versions of CUDA. In case you need to use a function that has been deprecated, you'll need to manually wrap it and add it to `libcusparse_deprecated.jl`. Luckily NVIDIA has versioned documentation nowadays, so at least we can look those up.

amontoison commented:
The issue is  `Array(coo)`. Julia wants to convert it as a CuMatrix and it's not supported by CUDA toolkit <= v"11.0".

amontoison committed:
Update test/cusparse/conversions.jl

maleadt commented on test/cusparse/conversions.jl:
```diff
@@ -70,3 +76,75 @@ end
 end
+
+if CUSPARSE.version() >= v"11.3"
+    @testset "conversions" begin
```
a testset "conversions" isn't really descriptive here -- the parent testset (derived from the filename) is already "conversions"

amontoison replied:
What do you think of `conversions between CuSparseMatrices`?

maleadt commented:
> The issue is Array(coo). Julia wants to convert it as a CuMatrix and it's not supported by CUDA toolkit <= v"11.0".

That seems like important functionality to disable, esp. seeing how the other PR relies on these conversions? Can you go through CSR ([cusparse<t>coo2csr()](https://docs.nvidia.com/cuda/archive/10.0/cusparse/index.html#cusparse-lt-t-gt-coo2csr) is available on earlier versions)?

amontoison commented:
I didn't disabled any functionality. I just added tests.
I think that the issue is not coo -> csr but coo -> cumatrix.
It seems that this conversion is required for `Array(coo)`.

I recently added coo <-> cumatrix but I use new routines (https://github.com/JuliaGPU/CUDA.jl/blob/master/lib/cusparse/conversions.jl#L701-L707).
We probably need something similar to https://github.com/JuliaGPU/CUDA.jl/blob/master/lib/cusparse/conversions.jl#L604-L661 for coo matrices with CUDA toolkit <= v"11.0".

**Update**: Did you mean enable instead of disable?

maleadt commented:
> I didn't disabled any functionality. I just added tests.

I assumed that the conversion had been broken by a recent PR, so only adding tests for the supported case (`CUSPARSE.version() >= v"11.3"`) seemed like a bad thing. But you're suggesting that this conversion (`Array(::CuSparseMatrixCOO)`) never worked?

amontoison committed:
Add conversions for old versions of CUDA toolkit

amontoison commented:
I'm sorry for the confusion.

With this PR, if fixed a bug that I introduced in https://github.com/JuliaGPU/CUDA.jl/pull/1637, I permuted `cooColInd` and `csc.rowVal` [here](https://github.com/JuliaGPU/CUDA.jl/blob/master/lib/cusparse/conversions.jl#L526).
I didn't checked that the conversion COO <-> CSC was not tested.

I added a test for it in this PR and I also tested that conversions work as well with other CuSparseMatrix.
Some of them were not implemented (BSR <-> CSC) or only supported by recent CUDA toolkit (the CI failures are related to these ones).

I think that `(Array(::CuSparseMatrixCOO))` was not working until https://github.com/JuliaGPU/CUDA.jl/pull/1611 when I added CuMatrix(coo).
I suppose that `Array(coo)` try to convert it into a `CuMatrix` as an intermediate step.

I removed all `CUSPARSE.version() >= v"11.3"` and added conversions for old versions of CUDA toolkit.

amontoison committed:
Always sort COO matrices by row

maleadt commented:
Ah OK, that clears it up. LGTM, thanks!

CarloLucibello commented:
@maleadt when can we get a new release with all these cusparse goodies from @amontoison ?

maleadt commented:
I don't want to rush the release, because it's breaking, and I want to make sure the breaking changes are exactly how we want them to be (specifically, the new artifact loading mechanism).
